api/routes/websoc/lib.ts

```diff
@@ -0,0 +1,425 @@
+              },
+            },
+          },
+        }))
+    );
+  }
+
+  if (parsedQuery.fullCourses && parsedQuery.fullCourses !== "ANY") {
+    switch (parsedQuery.fullCourses) {
+      case "FullOnly":
+        AND.push({ sectionFull: true, waitlistFull: true });
+        break;
+      case "OverEnrolled":
+        AND.push({ overEnrolled: true });
+        break;
+      case "SkipFull":
+        AND.push({ sectionFull: true, waitlistFull: false });
+        break;
+      case "SkipFullWaitlist":
+        AND.push({ sectionFull: false, waitlistFull: false });
+    }
+  }
+
+  switch (parsedQuery.cancelledCourses) {
+    case undefined:
+    case "Exclude":
+      AND.push({ cancelled: false });
+      break;
+    case "Only":
+      AND.push({ cancelled: true });
+  }
+
+  if (parsedQuery.sectionCodes) {
+    OR.push(
+      ...parsedQuery.sectionCodes.map((code) => ({
+        sectionCode: code.includes("-")
+          ? {
+              gte: parseInt(code.split("-")[0]),
+              lte: parseInt(code.split("-")[1]),
+            }
+          : parseInt(code),
+      }))
+    );
+  }
+
+  if (parsedQuery.units) {
+    OR.push(
+      ...parsedQuery.units.map((u) => ({
+        units:
+          u === "VAR"
+            ? { contains: "-" }
+            : { startsWith: parseFloat(u).toString() },
+      }))
+    );
+  }
+
+  return {
+    AND,
+
+    /**
+     * if OR non-empty, then use it; otherwise undefined means "ignore this field"
+     */
+    OR: OR.length ? OR : undefined,
+  };
+}
+
+/**
+ * Normalize a parsed query into array of objects that can be passed to ``callWebSocAPI``.
+ *
+ * To support batch queries for ``units`` and ``sectionCodes``,
+ * copies of the normalized query are created for every ``units``
+ * argument specified and for every 5 ``sectionCodes`` argument specified.
+ * @param query Zod-parsed query object.
+ */
+export function normalizeQuery(query: Query): WebsocAPIOptions[] {
+  const {
+    units: _units,
+    sectionCodes: _sectionCodes,
+    ...baseQuery
+  } = {
+    ...query,
+    instructorName: query.instructorName ?? "",
+    sectionCodes: query.sectionCodes?.join(","),
+    building: query.building ?? "",
+    room: query.room ?? "",
+    courseNumber: query.courseNumber?.join(","),
+    days: query.days?.join(""),
+  };
+  if (query.units && query.sectionCodes) {
+    if (query.units.length === 1 && query.sectionCodes.length < 6) {
+      return [
+        {
+          ...baseQuery,
+          units: query.units[0],
+          sectionCodes: query.sectionCodes.join(","),
+        },
+      ];
+    }
+
+    const keys = query.sectionCodes
+      .map((_, i) => (i % 5 === 0 ? i : null))
+      .filter(notNull);
+
+    return query.units
+      .map((units) => ({ ...baseQuery, units }))
+      .flatMap((copiedQuery) =>
+        keys.map((k) => ({
+          ...copiedQuery,
+          sectionCodes: (query.sectionCodes ?? [])
+            .slice(k * 5, (k + 1) * 5)
+            .join(","),
```

Comment:
```suggestion
          sectionCodes: query.sectionCodes?
            .slice(k * 5, (k + 1) * 5)
            .join(",") || [],
```
I generally don't like doing an early `or` because I could negate short circuiting. e.g. If `query.SectionCodes` is undefined, then it can immediately short circuit to `[]`, whereas creating the array early will result in it being sliced and joined.